### PR TITLE
SWC-7064, SWC-7092 - Use Elemental2 for JsInterop models

### PR DESCRIPTION
**Pull request by nickgros**
To support SWC-7064, SWC-7092, we will have to update logic related to File uploads. Instead of manipulating JSNI code, I opted to 
* pull in Elemental2, which provides JsInterop-compatible models for native JavaScript functionality (such as FileList, File, Blob)
* Replace our custom models with Elemental2 implementations (JSON, JsObject, Promise)
* Remove some basic JSNI code related to Blobs that could be replaced with basic JsInterop code

### CODE REVIEW

File: src/main/java/org/sagebionetworks/web/client/PortalGinModule.java
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/PortalGinModule.java:1]: ...most of the changes here are imports

File: src/main/java/org/sagebionetworks/web/client/SynapseJSNIUtils.java [hunk 1]
```diff
@@ -43,10 +43,6 @@ public LayoutResult nChartlayout(
 
   public void setPageDescription(String newDescription);
 
-  public JavaScriptObject getFileList(String fileFieldId);
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/SynapseJSNIUtils.java:46]: Moved these to a new `SynapseJsInteropUtils` interface & class

File: src/main/java/org/sagebionetworks/web/client/SynapseJsInteropUtilsImpl.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+import com.google.gwt.dom.client.Element;
+import elemental2.dom.Blob;
+import elemental2.dom.DomGlobal;
+import elemental2.dom.File;
+import elemental2.dom.FileList;
+import elemental2.dom.HTMLInputElement;
+import jsinterop.base.Js;
+import org.sagebionetworks.web.client.callback.MD5Callback;
+
+public class SynapseJsInteropUtilsImpl implements SynapseJsInteropUtils {
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/SynapseJsInteropUtilsImpl.java:13]: Rewrote all of these from JSNI to JsInterop using the Elemental2 classes

File: src/main/java/org/sagebionetworks/web/client/SynapseJsInteropUtilsImpl.java [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    if (unSplitNames.isEmpty()) return null;
+    return unSplitNames.split(";");
+  }
+
+  @Override
+  public String getWebkitRelativePath(FileList fileList, double index) {
+    return (String) Js
+      .asPropertyMap(fileList.item(index))
+      .get("webkitRelativePath");
+  }
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/SynapseJsInteropUtilsImpl.java:48]: Elemental2's `File` model does not surface the `webkitRelativePath` field, but we can easily get it with `Js.asPropertyMap(obj).get("key")`

File: src/main/java/org/sagebionetworks/web/client/widget/aws/AwsSdk.java [hunk 4]
```diff
@@ -62,43 +63,43 @@ public void upload(
...
+  				console.log("Upload error", err.code, err.message);
+  				upload.abort();
+  				callback.@org.sagebionetworks.web.client.widget.upload.S3DirectUploadHandler::uploadFailed(Ljava/lang/String;)(err.message);
+  			} else {
+  				//success
+  				callback.@org.sagebionetworks.web.client.widget.upload.S3DirectUploadHandler::uploadSuccess()();
+  			}
+  		};
+  		upload.send(listener);
+  	}-*/;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/widget/aws/AwsSdk.java]:
```suggestion
		var parameters = {
			Key : key,
			ContentType : contentType,
			Body : file
		};
		var upload = new $wnd.AWS.S3.ManagedUpload({
			//			partSize: 10 * 1024 * 1024,  //default 5MB
			//			queueSize: 1, // default 4
			leavePartsOnError : true,
			params : parameters,
			service : s3
		});
		upload
				.on(
						'httpUploadProgress',
						function(evt) {
							// report progress
							callback.@org.sagebionetworks.web.client.widget.upload.S3DirectUploadHandler::updateProgress(D)(evt.loaded/evt.total);
						});
		var listener = function(err, data) {
			if (err) {
				//error
				console.log("Upload error", err.code, err.message);
				upload.abort();
				callback.@org.sagebionetworks.web.client.widget.upload.S3DirectUploadHandler::uploadFailed(Ljava/lang/String;)(err.message);
			} else {
				//success
				callback.@org.sagebionetworks.web.client.widget.upload.S3DirectUploadHandler::uploadSuccess()();
			}
		};
		upload.send(listener);
	}-*/;
```

File: src/main/java/org/sagebionetworks/web/client/SynapseJSNIUtilsImpl.java [hunk 5]
```diff
@@ -404,26 +322,26 @@ private static final native double _getLastModified(
...
+  			  $wnd.calculateFileMd5Worker = new $wnd.Worker("/workers/calculateFileMd5Worker.js");
+  			};
+  			$wnd.calculateFileMd5Worker.onmessage = function(event) {
+  				md5Callback.@org.sagebionetworks.web.client.callback.MD5Callback::setMD5(Ljava/lang/String;)(event.data);
+  			};
+  			$wnd.calculateFileMd5Worker.postMessage(file);
+  	  } else {
+  			md5Callback.@org.sagebionetworks.web.client.callback.MD5Callback::setMD5(Ljava/lang/String;)(null);
+    	}
+  	}-*/;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/SynapseJSNIUtilsImpl.java]:
```suggestion
		if ($wnd.Worker) {
			if (!$wnd.calculateFileMd5Worker) {
			  $wnd.calculateFileMd5Worker = new $wnd.Worker("/workers/calculateFileMd5Worker.js");
			};
			$wnd.calculateFileMd5Worker.onmessage = function(event) {
				md5Callback.@org.sagebionetworks.web.client.callback.MD5Callback::setMD5(Ljava/lang/String;)(event.data);
			};
			$wnd.calculateFileMd5Worker.postMessage(file);
	  } else {
			md5Callback.@org.sagebionetworks.web.client.callback.MD5Callback::setMD5(Ljava/lang/String;)(null);
  	}
	}-*/;
```

File: src/main/java/org/sagebionetworks/web/client/GlobalApplicationStateImpl.java [hunk 6]
```diff
@@ -494,53 +494,53 @@ private static final native void _initializeDragDrop(
...
+  			//if files are dropped into the root panel, then ignore the event (do not open file contents if user does not have the upload dialog open).
+  			rootPanel.addEventListener('drop', function(e) {
+  				e.preventDefault();
+  			});
+  			rootPanel.addEventListener('dragenter', allowDrag);
+  			rootPanel.addEventListener('dragover', allowDrag);
+  		} catch (err) {
+  			console.error(err);
+  		}
+  	}-*/;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/GlobalApplicationStateImpl.java]:
```suggestion
		try {
			function showDropZone() {
				dropZone.style.display = "block";
			}

			function hideDropZone() {
				dropZone.style.display = "none";
			}

			$wnd
					.addEventListener(
							'dragenter',
							function(e) {
								if (globalAppState.@org.sagebionetworks.web.client.GlobalApplicationStateImpl::isDragAndDropListenerSet()()) {
									showDropZone();
								}
							});

			function allowDrag(e) {
				e.dataTransfer.dropEffect = 'copy';
				e.preventDefault();
			}

			function handleDrop(e) {
				e.preventDefault();
				hideDropZone();
  				globalAppState.@org.sagebionetworks.web.client.GlobalApplicationStateImpl::onDrop(Lelemental2/dom/FileList;)(e.dataTransfer.files);
			}

			dropZone.addEventListener('dragenter', allowDrag);
			dropZone.addEventListener('dragover', allowDrag);

			dropZone.addEventListener('drop', handleDrop);

			//if files are dropped into the root panel, then ignore the event (do not open file contents if user does not have the upload dialog open).
			rootPanel.addEventListener('drop', function(e) {
				e.preventDefault();
			});
			rootPanel.addEventListener('dragenter', allowDrag);
			rootPanel.addEventListener('dragover', allowDrag);
		} catch (err) {
			console.error(err);
		}
	}-*/;
```

File: src/main/java/org/sagebionetworks/web/client/widget/upload/CroppedImageUploadViewImpl.java [hunk 7]
```diff
@@ -184,33 +184,33 @@ private static native void _loadImage(
...
+  						circle : true
+  					})
+  					.then(
+  							function(blob) {
+  								v.@org.sagebionetworks.web.client.widget.upload.CroppedImageUploadViewImpl::saveCroppedImage(Lelemental2/dom/File;)(blob);
+  							});
+  		} catch (err) {
+  			console.error(err);
+  		}
+  	}-*/;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/widget/upload/CroppedImageUploadViewImpl.java]:
```suggestion
		try {
			$wnd.cropping
					.result({
						type : 'blob',
						format : 'png',
						quality : 1,
						circle : true
					})
					.then(
							function(blob) {
								v.@org.sagebionetworks.web.client.widget.upload.CroppedImageUploadViewImpl::saveCroppedImage(Lelemental2/dom/File;)(blob);
							});
		} catch (err) {
			console.error(err);
		}
	}-*/;
```

File: src/main/java/org/sagebionetworks/web/client/widget/upload/ImageUploadViewImpl.java [hunk 8]
```diff
@@ -143,75 +143,72 @@ private static native void _loadImage(
...
-		console.log('pica resizer initialized');
-	}-*/;
-
-  public void resizeComplete(JavaScriptObject blob) {
+      console.log('initializing pica resizer');
+      $wnd.resizer = $wnd.pica({
+        features: ['js', 'wasm', 'ww']
+      });
+      console.log('pica resizer initialized');
+    }-*/;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/widget/upload/ImageUploadViewImpl.java]:
```suggestion
		console.log('initializing pica resizer');
		$wnd.resizer = $wnd.pica({
			features : [ 'js', 'wasm', 'ww' ]
		});
		console.log('pica resizer initialized');
	}-*/;


```

File: src/main/java/org/sagebionetworks/web/client/GlobalApplicationStateImpl.java [hunk 9]
```diff
@@ -520,7 +520,7 @@ function allowDrag(e) {
 			function handleDrop(e) {
 				e.preventDefault();
 				hideDropZone();
-				globalAppState.@org.sagebionetworks.web.client.GlobalApplicationStateImpl::onDrop(Lcom/google/gwt/core/client/JavaScriptObject;)(e.dataTransfer.files);
+  				globalAppState.@org.sagebionetworks.web.client.GlobalApplicationStateImpl::onDrop(Lelemental2/dom/FileList;)(e.dataTransfer.files);
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/GlobalApplicationStateImpl.java]:
```suggestion
				globalAppState.@org.sagebionetworks.web.client.GlobalApplicationStateImpl::onDrop(Lelemental2/dom/FileList;)(e.dataTransfer.files);
```

File: src/main/java/org/sagebionetworks/web/client/widget/upload/ImageUploadViewImpl.java [hunk 10]
```diff
@@ -143,58 +143,55 @@ private static native void _loadImage(
...
+              })
+        }
+        imgElement.addEventListener('load', onImageLoad, false)
+        if (file && canResize) {
+          imgElement.src = $wnd.URL.createObjectURL(file)
+        } else {
+          // send back original content
+          v.@org.sagebionetworks.web.client.widget.upload.ImageUploadViewImpl::noResizeNecessary(Lelemental2/dom/Blob;)(file)
+        }
+      }-*/;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/widget/upload/ImageUploadViewImpl.java]:
```suggestion
																																																																							var fileToUploadElement = $doc.getElementById(fileFieldId);
																																																																							var file;
																																																																							var canResize = true;
																																																																							if (fileToUploadElement && 'files' in fileToUploadElement) {
																																																																							file = fileToUploadElement.files[0];
																																																																							var ext = file.name.split('.').pop();
																																																																							// resize if it is a supported file type, and the file size > 3MB
																																																																							canResize = [ 'bmp', 'jpg', 'jpeg', 'png' ].indexOf(ext) > -1 && file.size > 3145728;
																																																																							}
																																																																							
																																																																							var imgElement = $doc.createElement('img');
																																																																							var onImageLoad = function() {
																																																																							// Create an empty canvas element of the same dimensions as the original
																																																																							originalCanvas.width = imgElement.width;
																																																																							originalCanvas.height = imgElement.height;
																																																																							// Copy the image contents to the canvas
																																																																							var ctx = originalCanvas.getContext("2d");
																																																																							ctx.drawImage(imgElement, 0, 0);
																																																																							var maxWidth = 2048;
																																																																							var maxHeight = 2048;
																																																																							if (imgElement.width > maxWidth) {
																																																																							// continue resize based on width
																																																																							var ratio = maxWidth / imgElement.width; // get ratio for scaling image
																																																																							resizedCanvas.width = maxWidth;
																																																																							resizedCanvas.height = imgElement.height * ratio;
																																																																							} else {
																																																																							// continue resize based on height
																																																																							var ratio = maxHeight / imgElement.height; // get ratio for scaling image
																																																																							resizedCanvas.height = maxHeight;
																																																																							resizedCanvas.width = imgElement.width * ratio;
																																																																							}
																																																																							
																																																																							// Resize & convert to blob
																																																																							$wnd.resizer.resize(originalCanvas, resizedCanvas)
																																																																							.then (
																																																																							function(result) {
																																																																							$wnd.resizer.toBlob(result, 'image/jpeg', 90)
																																																																							.then(
																																																																							function(blob) {
																																																																							v.@org.sagebionetworks.web.client.widget.upload.ImageUploadViewImpl::resizeComplete(Lelemental2/dom/Blob;)(blob);
																																																																							});
																																																																							});
																																																																							};
																																																																							imgElement.addEventListener('load', onImageLoad, false);
																																																																							
																																																																							if (file && canResize) {
																																																																							imgElement.src = $wnd.URL.createObjectURL(file);
																																																																							} else {
																																																																							// send back original content
																																																																							v.@org.sagebionetworks.web.client.widget.upload.ImageUploadViewImpl::noResizeNecessary(Lelemental2/dom/Blob;)(file);
																																																																							}
																																																																							}-*/;
```

File: src/main/java/org/sagebionetworks/web/client/widget/entity/download/Uploader.java [hunk 11]
```diff
@@ -564,11 +570,8 @@ public void setCurrentExternalUploadUrl(String currentExternalUploadUrl) {
   }
 
   public void checkFileSize() throws IllegalArgumentException {
-    JavaScriptObject blob = synapseJsniUtils.getFileBlob(
-      this.currIndex,
-      fileList
-    );
-    long fileSize = (long) synapseJsniUtils.getFileSize(blob);
+    File blob = fileList.item(this.currIndex);
+    long fileSize = blob.size;
```
Comment (nickgros) [inline src/main/java/org/sagebionetworks/web/client/widget/entity/download/Uploader.java:574]: In many cases where it was more simple, I just replaced calls on utils methods with direct calls to the new JsInterop methods

File: pom.xml [hunk 12]
```diff
@@ -950,7 +950,19 @@
...
+    <dependency>
+      <groupId>com.google.elemental2</groupId>
+      <artifactId>elemental2-core</artifactId>
+      <version>1.2.2</version>
+    </dependency>
+
+    <dependency>
+      <groupId>com.google.elemental2</groupId>
+      <artifactId>elemental2-dom</artifactId>
+      <version>1.2.2</version>
```
Comment (nickgros) [inline pom.xml:965]: [Elemental2](https://github.com/google/elemental2) provides JsInterop bindings for browser APIs.

**Review by jay-hodgson — APPROVED** [review]

**Pull request merged**